File: 290__Word_Pattern.py
class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        dic_pattern_str = {}
        dic_str_pattern = {}
        s = str.split()
        if len(s) != len(pattern):
            return False
        for i, char in enumerate(pattern):
            if s[i] in dic_str_pattern:
                if dic_str_pattern[s[i]] != char:
                    return False
            else:
                dic_str_pattern[s[i]] = char
            if char in dic_pattern_str:
                if dic_pattern_str[char] != s[i]:
                    return False
            else:
                dic_pattern_str[char] = s[i]
        return True

File: Java/290__Word_Pattern.java
class Solution {
    public boolean wordPattern(String pattern, String str) {
        String[] str_arr = str.split(" ");
        System.out.println(str_arr.length);
        if (str_arr.length != pattern.length())
            return false;
        HashMap<Character, String> map = new HashMap<>();
        for (int i = 0; i < pattern.length(); i++) {
            char c = pattern.charAt(i);
            if (map.containsKey(c)) {
                if (!map.get(c).equals(str_arr[i]))
                    return false;
            } else {
                if (map.containsValue(str_arr[i]))
                    return false;
                map.put(c, str_arr[i]);
            }
        }
        return true;
    }
}
